packages/core/src/extractors/MetadataExtractors.ts

```diff
@@ -0,0 +1,400 @@
+  llm: LLM;
+
+  /**
+   * Can work for mixture of text and non-text nodes
+   * @type {boolean}
+   * @default false
+   */
+  isTextNodeOnly: boolean = false;
+
+  /**
+   * Number of nodes to extrct titles from.
+   * @type {number}
+   * @default 5
+   */
+  nodes: number = 5;
+
+  /**
+   * The prompt template to use for the title extractor.
+   * @type {string}
+   * @default "{title}"
+   */
+  nodeTemplate: string = defaultTitleExtractorPromptTemplate({
+    context_str: "{title}",
+  });
+
+  /**
+   * The prompt template to merge title with..
+   * @type {string}
+   * @default "{title}"
+   */
+  combineTemplate: string = defaultTitleCombinePromptTemplate({
+    context_str: "{title}",
+  });
+
+  /**
+   * Constructor for the TitleExtractor class.
+   * @param {LLM} llm LLM instance.
+   * @param {number} nodes Number of nodes to extract titles from.
+   * @param {string} node_template The prompt template to use for the title extractor.
+   * @param {string} combine_template The prompt template to merge title with..
+   */
+  constructor(
+    llm: LLM,
+    nodes: number = 5,
+    node_template: string = defaultTitleExtractorPromptTemplate({
+      context_str: "{title}",
+    }),
+    combine_template: string = defaultTitleCombinePromptTemplate({
+      context_str: "{title}",
+    }),
+  ) {
+    super();
+
+    this.llm = llm;
+    this.nodes = nodes;
+    this.nodeTemplate = node_template;
+    this.combineTemplate = combine_template;
+  }
+
+  /**
+   * Extract titles from a list of nodes.
+   * @param {BaseNode[]} nodes Nodes to extract titles from.
+   * @returns {Promise<Record<string, any>[]>} Titles extracted from the nodes.
+   */
+  async extract(nodes: BaseNode[]): Promise<Record<string, any>[]> {
+    const nodesToExtractTitle: BaseNode[] = [];
+
+    for (let i = 0; i < this.nodes; i++) {
+      if (nodesToExtractTitle.length >= nodes.length) break;
+
+      if (this.isTextNodeOnly && !(nodes[i] instanceof TextNode)) continue;
+
+      nodesToExtractTitle.push(nodes[i]);
+    }
+
+    if (nodesToExtractTitle.length === 0) return [];
+
+    let titlesCandidates: string[] = [];
+    let title: string = "";
+
+    for (let i = 0; i < nodesToExtractTitle.length; i++) {
+      const completion = await this.llm.complete({
+        prompt: defaultTitleExtractorPromptTemplate({
+          context_str: nodesToExtractTitle[i].getContent(MetadataMode.ALL),
+        }),
+      });
+
+      titlesCandidates.push(completion.text);
+    }
+
+    if (nodesToExtractTitle.length > 1) {
+      const combinedTitles = titlesCandidates.join(",");
+
+      const completion = await this.llm.complete({
+        prompt: defaultTitleCombinePromptTemplate({
+          context_str: combinedTitles,
+        }),
+      });
+
+      title = completion.text;
+    }
+
+    if (nodesToExtractTitle.length === 1) {
+      title = titlesCandidates[0];
+    }
+
+    return nodes.map((_) => ({
+      documentTitle: title.trim().replace(STRIP_REGEX, ""),
```

Comment:
What exactly is TitleExtractor supposed to do? I see here we're using the same title for all of the nodes?